contracts/BondCalculator.sol

```diff
@@ -0,0 +1,232 @@
+    // This number is bound by the total number of possible components and agents
+    uint64 targetNewUnits;
+    // DAO set weight factors
+    // The sum of factors cannot exceed the value of 10_000 (100% with a 0.01% step)
+    uint16[4] weightFactors;
+}
+
+// The size of the struct is 160 + 32 + 160 + 96 = 256 + 192 (2 slots)
+struct Product {
+    // priceLP (reserve0 / totalSupply or reserve1 / totalSupply) with 18 additional decimals
+    // priceLP = 2 * r0/L * 10^18 = 2*r0*10^18/sqrt(r0*r1) ~= 61 + 96 - sqrt(96 * 112) ~= 53 bits (if LP is balanced)
+    // or 2* r0/sqrt(r0) * 10^18 => 87 bits + 60 bits = 147 bits (if LP is unbalanced)
+    uint160 priceLP;
+    // Supply of remaining OLAS tokens
+    // After 10 years, the OLAS inflation rate is 2% per year. It would take 220+ years to reach 2^96 - 1
+    uint96 supply;
+    // Token to accept as a payment
+    address token;
+    // Current OLAS payout
+    // This value is bound by the initial total supply
+    uint96 payout;
+    // Max bond vesting time
+    // 2^32 - 1 is enough to count 136 years starting from the year of 1970. This counter is safe until the year of 2106
+    uint32 vesting;
+}
+
+/// @title BondCalculator - Smart contract for bond calculation payout in exchange for OLAS tokens based on dynamic IDF.
+/// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
+/// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
+/// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
+contract BondCalculator is GenericBondCalculator {
+    event OwnerUpdated(address indexed owner);
+    event DiscountParamsUpdated(DiscountParams newDiscountParams);
+
+    // Maximum sum of discount factor weights
+    uint256 public constant MAX_SUM_WEIGHTS = 10_000;
+    // veOLAS contract address
+    address public immutable ve;
+
+    // Contract owner
+    address public owner;
+    // Discount params
+    DiscountParams public discountParams;
+
+
+    /// @dev Bond Calculator constructor.
+    /// @param _olas OLAS contract address.
+    /// @param _tokenomics Tokenomics contract address.
+    /// @param _ve veOLAS contract address.
+    /// @param _discountParams Discount factor parameters.
+    constructor(address _olas, address _tokenomics, address _ve, DiscountParams memory _discountParams)
+        GenericBondCalculator(_olas, _tokenomics)
+    {
+        // Check for zero address
+        if (_ve == address(0)) {
+            revert ZeroAddress();
+        }
+
+        ve = _ve;
+        owner = msg.sender;
+
+        // Check for zero values
+        if (_discountParams.targetNewUnits == 0 || _discountParams.targetVotingPower == 0) {
+            revert ZeroValue();
+        }
+        // Check the sum of factors that cannot exceed the value of 10_000 (100% with a 0.01% step)
+        uint256 sumWeights;
+        for (uint256 i = 0; i < _discountParams.weightFactors.length; ++i) {
+            sumWeights += _discountParams.weightFactors[i];
```

Comment:
_discountParams.weightFactors[i] must > min and sumWeights > 0

Reply:
weights can all be zero, leaving as is.